EventGenerator/src/Pileup_module.cc

```diff
@@ -0,0 +1,180 @@
+#include "Mu2eUtilities/inc/stoppedMuonList.hh"
+#include "EventGenerator/inc/ParticleGeneratorTool.hh"
+
+namespace mu2e {
+
+  //================================================================
+  class Pileup : public art::EDProducer {
+  public:
+    struct Config {
+      using Name=fhicl::Name;
+      using Comment=fhicl::Comment;
+      fhicl::Atom<art::InputTag> inputSimParticles{Name("inputSimParticles"),
+          Comment("A SimParticleCollection with muons stopping in the target.")};
+      fhicl::Atom<std::string> stoppingTargetMaterial{
+        Name("stoppingTargetMaterial"),
+          Comment("Material determines muon life time, capture fraction, and particle spectra.\n"
+                  "Only aluminum (Al) is supported, emisson spectra for other materials are not implemented.\n"),
+          "Al" };
+
+      fhicl::DelegatedParameter captureProducts{Name("captureProducts"), Comment("A sequence of ParticleGenerator tools implementing capture products.")};
+      fhicl::DelegatedParameter decayProducts{Name("decayProducts"), Comment("A sequence of ParticleGenerator tools implementing decay products.")};
+
+      fhicl::Atom<unsigned> verbosity{Name("verbosity"),0};
+    };
+
+    using Parameters= art::EDProducer::Table<Config>;
+    explicit Pileup(const Parameters& conf);
+
+    virtual void produce(art::Event& event) override;
+
+    //----------------------------------------------------------------
+  private:
+    double muonLifeTime_;
+    double decayFraction_;
+
+    art::ProductToken<SimParticleCollection> const simsToken_;
+
+    unsigned verbosity_;
+
+    art::RandomNumberGenerator::base_engine_t& eng_;
+    CLHEP::RandFlat randFlat_;
+    CLHEP::RandExponential randExp_;
+
+    std::vector<std::unique_ptr<ParticleGeneratorTool>> muonDecayGenerators_;
+    std::vector<std::unique_ptr<ParticleGeneratorTool>> muonCaptureGenerators_;
+
+    void addParticles(StageParticleCollection* output, art::Ptr<SimParticle> mustop, double time, ParticleGeneratorTool* gen);
+  };
+
+  //================================================================
+  Pileup::Pileup(const Parameters& conf)
+    : EDProducer{conf}
+    , muonLifeTime_{GlobalConstantsHandle<PhysicsParams>()->getDecayTime(conf().stoppingTargetMaterial())}
+    , decayFraction_{GlobalConstantsHandle<PhysicsParams>()->getDecayFraction(conf().stoppingTargetMaterial())}
+    , simsToken_{consumes<SimParticleCollection>(conf().inputSimParticles())}
+    , verbosity_{conf().verbosity()}
+    , eng_{createEngine(art::ServiceHandle<SeedService>()->getSeed())}
+    , randFlat_{eng_}
+    , randExp_{eng_}
+  {
+    produces<mu2e::StageParticleCollection>();
+    if(verbosity_ > 0) {
+      mf::LogInfo log("Pileup");
+      log<<"stoppingTargetMaterial = "<<conf().stoppingTargetMaterial()
+         <<", muon lifetime = "<<muonLifeTime_
+         <<", decay fraction = "<<decayFraction_
+         <<std::endl;
+    }
+
+    if(conf().stoppingTargetMaterial() != "Al") {
```

Comment:
It is the individual tools that need to be configured for different materials.  Shouldn't the throw be in finishInitialization?  As far as what this module does (lifetime, decay fraction) many materials besides Al are supported.